homeassistant/components/zwave_js/discovery_data_template.py

```diff
@@ -251,3 +251,102 @@ def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]
     def current_tilt_value(resolved_data: dict[str, Any]) -> ZwaveValue | None:
         """Get current tilt ZwaveValue from resolved data."""
         return resolved_data["tilt_value"]
+
+
+@dataclass
+class FanSpeedDataTemplate:
+    """Mixin to define get_speed_config."""
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """
+        Get the fan speed configuration for this device.
+
+        Values should indicate the highest allowed device setting for each
+        actual speed.
+        """
+        # pylint: disable=no-self-use
+        return None
+
+
+@dataclass
+class FixedFanSpeedDataTemplate(BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate):
+    """
+    Specifies a fixed set of fan speeds.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=FixedFanSpeedDataTemplate(
+              speeds=[32,65,99]
+          ),
+      ),
+
+    `speeds` indicates the maximum setting on the underlying fan controller
+    for each actual speed.
+    """
+
+    speeds: list[int] = field(default_factory=list)
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """Get the fan speed configuration for this device."""
+        return self.speeds
+
+
+@dataclass
+class ConfigurableFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate
+):
+    """
+    Gets fan speeds based on a configuration value.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=ConfigurableFanSpeedDataTemplate(
+            configuration_option=ZwaveValueID(
+                5, CommandClass.CONFIGURATION, endpoint=0
+            ),
+            configuration_value_to_speeds={0: [32, 65, 99], 1: [24, 49, 74, 99]},
+          ),
+      ),
+
+    `configuration_option` is a reference to the setting that determines how
+    many speeds are supported.
+
+    `configuration_value_to_speeds` maps the values from `configuration_option`
+    to a list of speeds.  The specified speeds indicate the maximum setting on
+    the underlying switch for each actual speed.
+    """
+
+    configuration_option: ZwaveValueID | None = None
```

Comment:
If this attribute is required we can move it to the mixin `FanSpeedDataTemplate`.

Reply:
My idea for `FanSpeedDataTemplate` is that it'd be the interface that `ConfiguredSpeedRangeZwaveFan` uses to get information about what speeds a fan supports in a generic form,  i.e., so that `ConfiguredSpeedRangeZwaveFan` doesn't need to know that a fan was configured with `ConfiguredSpeedRangeZwaveFan` (for HomeSeer controllers) or `FixedFanSpeedDataTemplate` (probably most other zwave fan controllers).

This particular attribute is only needed by this specific subclass of `FanSpeedDataTemplate`, and it's of no interest to `ConfiguredSpeedRangeZwaveFan`.

Reply:
Ok, then let's add another mixin for this subclass so we can mark this attribute as required for this subclass.

Reply:
Done (and same for FixedFanSpeedDataTemplate)